### PR TITLE
Peer delete did not trigger ZEBRA_NEXTHOP_UNREGISTER to ZEBRA results…

### PR DESCRIPTION
**Pull request by vishaldhingra**
… in leak of route node objects in zebra

  RCA :

  bgp_nexthop_cache is allocated and ZEBRA_NEXTHOP_REGISTER is sent to ZEBRA.

  The association of BGP is as below :
  (nexthop_cache_table)<---->bgp_nexthop_cache---->peer
                                  |<--------------> bgp_info

  for peer association bnc has nht_info
  for bgp_info bnc has list count

  Allocation of bnc :
  Scenario 1: When peer pointer exists
  Scenario 2: When peer pointer does not exists and bgp_info exists via bgp_update in IBGP session, when next-hop is
              not self.

  Allocation is fine in all scenarios

  Deletion of bnc:
  Scenario 1: When peer pointer exists
  Scenario 2: When peer pointer does not exists

  The deletion is triggered via bgp_unlink_nexthop_by_peer()
  In this context bgp->nht_info sets to NULL, which is correct as peer is going to delete.
  but in bgp_unlink_nexthop_check() expects the nht_info should not be NULL.
  So it will not call bgp_unlink_nexthop_check to ZEBRA

  Fix :

  When bgp_unlink_nexthop_check() gets called;
  In both the above scenarios the PEER should be NULL and bgp_info list count should be zero.

  Also there was one issue : 2198

  The code checked-in in this context was added to avoid the crash, but it seems the crash was coming due to some other problem.

  I have tried topotest multiple times but did not face any crash. So I am reverting this change.

  If still it crashes then i will check that.

  Signed-off-by: vishaldhingra vdhingra@vmware.com

### CODE REVIEW

**Review by kssoman — APPROVED** [review]

File: bgpd/bgp_nht.c [hunk 1]
```diff
@@ -88,7 +88,7 @@ int bgp_find_nexthop(struct bgp_info *path, int connected)
 
 static void bgp_unlink_nexthop_check(struct bgp_nexthop_cache *bnc)
 {
-	if (LIST_EMPTY(&(bnc->paths)) && bnc->nht_info) {
+	if (LIST_EMPTY(&(bnc->paths)) && !bnc->nht_info) {
```
Comment (srimohans) [inline bgpd/bgp_nht.c:91]: So the only Fix done was to revert the changes done as part of #2198 ?
Did you talk to original author of the change for #2198 ?

Comment (srimohans) [inline bgpd/bgp_nht.c:91]: So the only Fix done was to revert the changes done as part of #2198 ?
Did you talk to original author of the change for #2198 ?

Comment (vishaldhingra) [reply bgpd/bgp_nht.c:91]: Yes, I have checked with lou Berger, here is the reply.

lberger [7:44 PM]
hi I'm in meetings this AM
crash was discussed in 2184 - submit your PR, maybe something has changed, ie.e., someone else put in the real fix -- we'll see if crash shows back up
